File: demo/insecure-code/secure_payment.py
# SECURE VERSION — the correct implementations for each vulnerability in vulnerable_payment.py.
# This is what the code looks like AFTER CodeQL and Snyk findings are remediated.
#
# Each function documents the remediation technique so it can be explained at interview.

import sqlite3
import os
import hashlib
import secrets
import json
import subprocess
import re


# ─── CWE-798 remediation: environment variables, never source code ────────────
# Credentials are injected at runtime via K8s Secrets (see k8s/rbac/rbac.yaml)
# or GCP Workload Identity (see terraform/gke.tf). They are never committed to git.
# CI pipelines use GitHub Actions secrets (repo → Settings → Secrets).
# Rotation is handled by KMS (terraform/kms.tf) without touching source code.
PAYMENT_GATEWAY_KEY = os.environ["PAYMENT_GATEWAY_KEY"]    # SAFE — sourced from env


# ─── CWE-89 remediation: parameterised query ─────────────────────────────────
# The '?' placeholder is a bound parameter. The database driver handles all escaping.
# User input is NEVER concatenated into the query string — it is passed separately.
# No SQL injection is possible regardless of what user_id contains.
def get_payment_details(user_id: str):
    conn = sqlite3.connect("payments.db")
    cursor = conn.cursor()
    query = "SELECT card_number, amount FROM payments WHERE user_id = ?"   # SAFE
    cursor.execute(query, (user_id,))   # user_id is a bind parameter, not SQL
    return cursor.fetchall()


# ─── CWE-78 remediation: allowlist + subprocess with list args ───────────────
# 1. report_name is validated against a strict allowlist (only a-z, 0-9, _ -)
#    Any payload like "jan; curl ..." is rejected before it reaches the shell.
# 2. subprocess.run() with a list (not a string) never invokes a shell at all.
#    Even if the allowlist had a gap, shell metacharacters are inert as list elements.
def generate_payment_report(report_name: str):
    if not re.fullmatch(r"[a-z0-9_-]{1,64}", report_name):
        raise ValueError(f"Invalid report name: {report_name!r}")
    subprocess.run(["generate_report.sh", report_name], check=True, shell=False)   # SAFE


# ─── CWE-327 remediation: PBKDF2-HMAC-SHA256 ────────────────────────────────
# PBKDF2 with a unique random salt and 600,000 iterations meets:
#   NIST SP 800-132 (key derivation for password storage)
#   PCI-DSS Requirement 3.5 (strong cryptography for stored sensitive data)
# The salt ensures two identical card numbers produce different hashes.
# 600,000 iterations makes brute-force attacks computationally infeasible.
def mask_card_number(card_number: str) -> str:
    salt = secrets.token_bytes(32)
    dk = hashlib.pbkdf2_hmac("sha256", card_number.encode(), salt, 600_000)
    return salt.hex() + ":" + dk.hex()   # SAFE — salt stored alongside the hash


# ─── CWE-502 remediation: JSON with schema validation ────────────────────────
# json.loads() has no code execution surface — it only parses data types.
# The payload is then validated against required keys before any field is trusted.
# If the session store needs richer types, use a typed schema validator (e.g. pydantic).
def restore_payment_session(session_blob: bytes) -> dict:
    session = json.loads(session_blob)   # SAFE — no arbitrary code execution
    required_keys = {"user_id", "expires_at", "cart_total_pence"}
    if not required_keys.issubset(session):
        raise ValueError(f"Malformed session — missing keys: {required_keys - session.keys()}")
    return session

File: demo/insecure-code/secure_payment.php
<?php
// SECURE VERSION — correct PHP for each vulnerability in vulnerable_payment.php.
// Patterns follow the OWASP PHP Security Cheat Sheet and PCI-DSS DSS v4.0.
// declare(strict_types=1) enforces type safety at the engine level.
declare(strict_types=1);


// ─── CWE-798 remediation: environment variables, never source code ────────────
// Credentials are injected at runtime via K8s Secrets (k8s/rbac/rbac.yaml)
// or GCP Workload Identity (terraform/gke.tf). The application code never
// contains or even knows credential values — it only knows the env var name.
// If the env var is missing, we fail loudly (misconfiguration is caught at startup).
$db_password = getenv('DB_PASSWORD')
    ?: throw new RuntimeException('DB_PASSWORD environment variable is not set');
$api_key = getenv('PAYMENT_GATEWAY_KEY')
    ?: throw new RuntimeException('PAYMENT_GATEWAY_KEY environment variable is not set');


// ─── CWE-89 remediation: PDO prepared statement ──────────────────────────────
// The :id placeholder is a bind parameter — no user input ever appears in SQL.
// Even a payload of "1 OR 1=1" is treated as a literal string value, not SQL.
// Type validation (ctype_digit) adds a second layer: non-numeric IDs are rejected
// before they reach the database driver at all.
function get_payment(PDO $pdo, string $payment_id): array
{
    if (!ctype_digit($payment_id)) {
        throw new InvalidArgumentException('payment_id must be a positive integer');
    }
    $stmt = $pdo->prepare('SELECT id, amount, status FROM payments WHERE id = :id');
    $stmt->execute([':id' => (int) $payment_id]);
    return $stmt->fetchAll(PDO::FETCH_ASSOC);  // SAFE
}


// ─── CWE-79 remediation: htmlspecialchars() output encoding ──────────────────
// htmlspecialchars() converts <, >, ", ', & to HTML entities.
// ENT_QUOTES ensures both single and double quotes are encoded (prevents attribute injection).
// Specifying 'UTF-8' explicitly prevents charset-based bypass attacks.
// For rich HTML output, use a template engine with auto-escaping (e.g. Twig).
function show_payment_status(): void
{
    $raw_status = $_GET['status'] ?? '';
    $status = htmlspecialchars($raw_status, ENT_QUOTES, 'UTF-8');  // SAFE
    echo "<div class='status'>Payment status: " . $status . "</div>";
}


// ─── CWE-78 remediation: allowlist + escapeshellarg() ────────────────────────
// Two-layer defence:
//   1. Regex allowlist: only accept decimal numbers with up to 2 decimal places.
//      A payload like "100; curl attacker.com" fails the regex and is rejected.
//   2. escapeshellarg(): wraps the value in single quotes for the shell, so even
//      if the allowlist had a gap, metacharacters cannot break out of the argument.
function generate_receipt(string $amount): void
{
    if (!preg_match('/^\d+(\.\d{1,2})?$/', $amount)) {
        throw new InvalidArgumentException('Invalid amount format: ' . $amount);
    }
    $safe_amount = escapeshellarg($amount);
    exec('generate_receipt.sh ' . $safe_amount, $output, $exit_code);  // SAFE
    if ($exit_code !== 0) {
        throw new RuntimeException('Receipt generation failed');
    }
}


// ─── CWE-22 remediation: allowlist map of templates ─────────────────────────
// The user can only request a name from a fixed list.
// The file path is never derived from user input — the map lookup selects
// a hardcoded server-side path. Path traversal is structurally impossible.
function load_template(string $name): void
{
    $allowed = [
        'success' => '/var/www/templates/payment-success.php',
        'failure' => '/var/www/templates/payment-failure.php',
        'pending' => '/var/www/templates/payment-pending.php',
    ];

    if (!array_key_exists($name, $allowed)) {
        throw new InvalidArgumentException("Unknown template name: {$name}");
    }

    include $allowed[$name];  // SAFE — path is a server-side constant
}


// ─── CWE-327 remediation: HMAC-SHA256 with a secret key ─────────────────────
// HMAC-SHA256 with a randomly generated server-side key meets PCI-DSS Req 3.4.
// The key is stored in an environment variable (never in code).
// HMAC is used (not plain SHA-256) because it resists length-extension attacks
// and requires the attacker to know the key to produce a valid hash.
// For passwords, prefer password_hash() with PASSWORD_ARGON2ID instead.
function hash_card_number(string $card_number): string
{
    $hmac_key = getenv('CARD_HASH_KEY')
        ?: throw new RuntimeException('CARD_HASH_KEY environment variable is not set');
    return hash_hmac('sha256', $card_number, $hmac_key);  // SAFE
}
